File: src/gwt/panmirror/src/editor/src/api/selection.ts
/*
 * selection.ts
 *
 * Copyright (C) 2020 by RStudio, PBC
 *
 * Unless you have received this program directly from RStudio pursuant
 * to the terms of a commercial license agreement with RStudio, then
 * this program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 */

import { Selection, NodeSelection } from 'prosemirror-state';
import { Schema } from 'prosemirror-model';
import { EditorView } from 'prosemirror-view';

import { NodeWithPos, setTextSelection } from 'prosemirror-utils';

import { kAddToHistoryTransaction, kRestoreLocationTransaction } from './transaction';

export function selectionIsWithin(selection: Selection, nodeWithPos: NodeWithPos) {
  const from = nodeWithPos.pos + 1;
  const to = from + nodeWithPos.node.nodeSize;
  return selectionIsWithinRange(selection, { from, to });
}

export function selectionIsWithinRange(selection: Selection, range: { from: number, to: number }) {
  return selection.anchor >= range.from && selection.anchor <= range.to;
}

export function selectionIsBodyTopLevel(selection: Selection) {
  const { $head } = selection;
  const parentNode = $head.node($head.depth - 1);
  return parentNode && parentNode.type === parentNode.type.schema.nodes.body;
}

export function selectionIsImageNode(schema: Schema, selection: Selection) {
  return selection instanceof NodeSelection && [schema.nodes.image, schema.nodes.figure].includes(selection.node.type);
}

export function selectionIsEmptyParagraph(schema: Schema, selection: Selection) {
  const { $head } = selection;
  return $head.parent.type === schema.nodes.paragraph && $head.parent.childCount === 0;
}

export function restoreSelection(view: EditorView, pos: number) {
  const tr = view.state.tr;
  if (pos < view.state.doc.nodeSize) {
    setTextSelection(pos)(tr);
    tr.setMeta(kAddToHistoryTransaction, false);
    tr.setMeta(kRestoreLocationTransaction, true);
    view.dispatch(tr);
  }
}

File: src/gwt/src/org/rstudio/core/client/dom/impl/Selection.java
/*
 * Selection.java
 *
 * Copyright (C) 2022 by Posit Software, PBC
 *
 * Unless you have received this program directly from Posit Software pursuant
 * to the terms of a commercial license agreement with Posit Software, then
 * this program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 */
package org.rstudio.core.client.dom.impl;

import org.rstudio.core.client.dom.NativeWindow;

import com.google.gwt.core.client.JavaScriptObject;
import com.google.gwt.dom.client.Node;

class Selection extends JavaScriptObject
{
   protected Selection()
   {
   }

   public static native Selection get() /*-{
      return $wnd.getSelection();
   }-*/;

   public static native Selection get(NativeWindow window) /*-{
      return window.getSelection();
   }-*/;

   public final native int getRangeCount() /*-{
      return this.rangeCount;
   }-*/;

   public final native Range getRangeAt(int index) /*-{
      return this.getRangeAt(index);
   }-*/;

   public final native void removeAllRanges() /*-{
      return this.removeAllRanges();
   }-*/;

   public final native void addRange(Range range) /*-{
      return this.addRange(range);
   }-*/;

   public final native Node getAnchorNode() /*-{
      return this.anchorNode;
   }-*/;

   public final native int getAnchorOffset() /*-{
      return this.anchorOffset;
   }-*/;

   public final native Node getFocusNode() /*-{
      return this.focusNode;
   }-*/;

   public final native int getFocusOffset() /*-{
      return this.focusOffset;
   }-*/;

   public final void setRange(Range selection)
   {
      removeAllRanges();
      if (selection != null)
         addRange(selection);
   }
}
